experimental/dds/tree2/src/codec/parameterized-dds.md

```diff
@@ -0,0 +1,252 @@
+# Parameterized DDSes
+
+This design document centers around the problem of providing better support for "parameterized DDSes,"
+aka a DDS which is configurable in some way by an application author.
+The current APIs are generally sufficient for DDSes whose only parameters affect runtime behavior in a way that doesn't have compatibility issues.
+For example, one could imagine writing a DDS with a "noisy" mode that emits many more diagnostic events on changes but otherwise behaves the same.
+There are no compatibility concerns with this sort of parameter:
+clients within a collaboration session are free to disagree on whether the extra events are emitted,
+and a document created by a client with one level of verbosity can be later opened by a client with a different level.
+
+However, the APIs start to have issues when considering parameters which might result in compatibility issues.
+There are a few reasons why supporting such parameters is desirable:
+
+1. The DDS supports functionality requiring persisted data on which applications may want different policies, or may not want at all (and would like pay-to-play semantics)
+    - Current examples: attribution, legacy `SharedTree`'s `summarizeHistory` flag
+    - Potential other examples: DDS implementations with configurable merge behavior
+2. The DDS has compatibility constraints which it would like to delegate to the application author to resolve. The motivating use case is:
+    - A DDS implements a new, more efficient snapshot or op format.
+      Package versions beneath the initial release of this snapshot encoding won't understand it and will crash if they attempt to read it.
+      The DDS author would like the application writer to select which format to use (and should provide documentation that code version Foo must be rolled out
+      to all clients before the new snapshot format is written).
```

Comment:
Why would the existing feature config not be used in these cases?

Reply:
I'm assuming you mean `IConfigProvider` here, let me know if that's not the case.

I think some Fluid customers might be OK using that, but in the general case it isn't really the right scope. That configuration is always provided by the host and is scoped to the entire container. We also don't persist it in a meaningfully way associated with the file, but we could fix that if we wanted to.

Since the config is supplied by the host, rolling out config changes which rely on features within the data model is awkward for ecosystems that separate the host and container author.

Since the config is scoped to the entire container, it's awkward or impossible to specify multiple configurations for different DDSes within a single container. There are cases where you might want this while rolling out a feature upgrade, but I'm imagining there are some where you'd want it in general. Attribution has been my go-to example here, but another one to consider is the two snapshot formats we have for merge-tree: matrix uses one and shared-string uses another due to perceived tradeoffs in encoding efficiency based on typical usage of the DDS.

Reply:
config is more about dark rollout, and testing in production. it is orthogonal to the need here. config in general should be short lived as well. if we need a longer lived setting it should go into our strongly types options object, like IContainerConfig, and have the full matrix of values tested. This is yet another mechanism to describe long term behavior, but rather than a knob to control the dds, it is more like dependency injection to provide new behaviors to the dds.